File: mac-app/QuartzTeachers/Scripting/TranscriptBuilder.swift
import Foundation

/// Turns raw terminal output into clean display lines.
///
/// The scripts produce terminal control sequences (colours, spinners that
/// redraw the same line with carriage returns, cursor movement). Teachers
/// should see tidy text, so this type strips ANSI escape sequences and
/// treats a carriage return as "start this line over", which collapses
/// spinner animation into its final state.
struct TranscriptBuilder {

    // MARK: - Stored properties

    /// Completed lines of output, oldest first.
    var lines: [String] = []

    /// The line currently being assembled (not yet ended with a newline).
    var currentLine: String = ""

    /// True when the previous chunk ended with a carriage return whose
    /// meaning (line ending vs. spinner redraw) depends on what follows.
    var hasPendingCarriageReturn: Bool = false

    // MARK: - Computed properties

    /// The full transcript as one display string.
    var displayText: String {
        var allLines: [String] = lines
        if !currentLine.isEmpty {
            allLines.append(currentLine)
        }
        return allLines.joined(separator: "\n")
    }

    // MARK: - Functions

    /// Feed a chunk of raw output from the PTY into the transcript.
    ///
    /// A pseudo-terminal turns every "\n" the script prints into "\r\n",
    /// so a carriage return followed by a newline is an ordinary line
    /// ending. Only a LONE carriage return is a spinner redrawing its
    /// line, which is when the current line restarts.
    mutating func append(rawText: String) {
        let cleaned: String = TranscriptBuilder.strippingControlSequences(from: rawText)
        // Work scalar-by-scalar: Swift groups "\r\n" into a SINGLE
        // Character (grapheme cluster), which would hide line endings.
        let newlineScalar: Unicode.Scalar = "\n"
        let carriageReturnScalar: Unicode.Scalar = "\r"
        for scalar in cleaned.unicodeScalars {
            if hasPendingCarriageReturn {
                hasPendingCarriageReturn = false
                if scalar == newlineScalar {
                    // "\r\n": a normal line ending.
                    lines.append(currentLine)
                    currentLine = ""
                    continue
                }
                // A lone "\r": spinner redraw — restart the line, then
                // process the current scalar normally below.
                currentLine = ""
            }
            if scalar == newlineScalar {
                lines.append(currentLine)
                currentLine = ""
            } else if scalar == carriageReturnScalar {
                hasPendingCarriageReturn = true
            } else {
                currentLine.unicodeScalars.append(scalar)
            }
        }
    }

    /// Removes ANSI escape sequences and stray control characters,
    /// keeping newlines, carriage returns, and tabs.
    ///
    /// Operates on Unicode scalars, not Characters, because Swift folds
    /// "\r\n" into one Character and would misclassify it as a control
    /// character to remove.
    static func strippingControlSequences(from text: String) -> String {
        var result: String = ""
        let scalars: [Unicode.Scalar] = Array(text.unicodeScalars)
        let escapeScalar: Unicode.Scalar = "\u{1B}"
        let bellScalar: Unicode.Scalar = "\u{07}"
        var index: Int = 0
        while index < scalars.count {
            let scalar: Unicode.Scalar = scalars[index]
            if scalar == escapeScalar {
                // Escape sequence: skip "ESC [ ... final-letter" (CSI) or
                // "ESC ] ... BEL" (OSC), or a single following scalar.
                let nextIndex: Int = index + 1
                if nextIndex < scalars.count && scalars[nextIndex] == "[" {
                    var scanIndex: Int = nextIndex + 1
                    while scanIndex < scalars.count {
                        let scanned: Unicode.Scalar = scalars[scanIndex]
                        let isFinalLetter: Bool = (scanned >= "A" && scanned <= "Z") || (scanned >= "a" && scanned <= "z")
                        if isFinalLetter {
                            break
                        }
                        scanIndex += 1
                    }
                    index = scanIndex + 1
                    continue
                }
                if nextIndex < scalars.count && scalars[nextIndex] == "]" {
                    var scanIndex: Int = nextIndex + 1
                    while scanIndex < scalars.count {
                        let scanned: Unicode.Scalar = scalars[scanIndex]
                        if scanned == bellScalar {
                            break
                        }
                        scanIndex += 1
                    }
                    index = scanIndex + 1
                    continue
                }
                index = index + 2
                continue
            }
            let isControl: Bool = scalar.value < 32
            let isKeeper: Bool = scalar == "\n" || scalar == "\r" || scalar == "\t"
            if isControl && !isKeeper {
                index += 1
                continue
            }
            result.unicodeScalars.append(scalar)
            index += 1
        }
        return result
    }
}

File: windows-app/Plantoir.Core/Scripting/TranscriptBuilder.cs
using System.Text;

namespace Plantoir.Core.Scripting;

/// <summary>
/// Turns raw terminal output into clean display lines.
///
/// A pseudo console converts every "\n" the script prints into "\r\n", so
/// CR followed by LF is an ordinary line ending. Only a LONE carriage
/// return is a spinner redrawing its line — it restarts the current line.
/// ANSI escape sequences (colours, cursor movement, OSC titles) are
/// stripped before the line machine runs.
/// </summary>
public sealed class TranscriptBuilder
{
    public const int MaximumRetainedLines = 4000;

    private readonly List<string> _lines = new();
    private readonly StringBuilder _currentLine = new();
    private bool _hasPendingCarriageReturn;
    private string? _cachedDisplayText;
    private long _version;

    /// <summary>Completed lines, oldest first (capped at MaximumRetainedLines).</summary>
    public IReadOnlyList<string> Lines => _lines;

    /// <summary>The unterminated line under construction.</summary>
    public string CurrentLine => _currentLine.ToString();

    /// <summary>Monotonic counter bumped on every append — cheap change detection.</summary>
    public long Version => _version;

    public void Append(string rawText)
    {
        _cachedDisplayText = null;
        _version++;
        string cleaned = StripControlSequences(rawText);
        foreach (char c in cleaned)
        {
            if (_hasPendingCarriageReturn)
            {
                _hasPendingCarriageReturn = false;
                if (c == '\n') { PushLine(); continue; }
                _currentLine.Clear();               // lone CR: spinner redraw
            }
            if (c == '\n') { PushLine(); }
            else if (c == '\r') { _hasPendingCarriageReturn = true; }
            else { _currentLine.Append(c); }
        }
    }

    private void PushLine()
    {
        _lines.Add(_currentLine.ToString());
        _currentLine.Clear();
        if (_lines.Count > MaximumRetainedLines)
            _lines.RemoveRange(0, _lines.Count - MaximumRetainedLines);
    }

    public string DisplayText
    {
        get
        {
            if (_cachedDisplayText is null)
            {
                var all = _currentLine.Length > 0
                    ? _lines.Append(_currentLine.ToString())
                    : _lines;
                _cachedDisplayText = string.Join("\n", all);
            }
            return _cachedDisplayText;
        }
    }

    /// <summary>The tail of the transcript, built without joining everything.</summary>
    public string RecentText(int maximumCharacters)
    {
        var collected = new List<string>();
        int budget = maximumCharacters;
        string current = _currentLine.ToString();
        if (current.Length > 0) { collected.Add(current); budget -= current.Length + 1; }
        for (int i = _lines.Count - 1; i >= 0 && budget > 0; i--)
        {
            collected.Add(_lines[i]);
            budget -= _lines[i].Length + 1;
        }
        collected.Reverse();
        return string.Join("\n", collected);
    }

    /// <summary>
    /// Strips ANSI CSI (ESC [ ... final letter), OSC (ESC ] ... BEL), other
    /// two-byte escapes, and control characters below 0x20 except \n \r \t.
    /// </summary>
    public static string StripControlSequences(string text)
    {
        var result = new StringBuilder(text.Length);
        int i = 0;
        while (i < text.Length)
        {
            char c = text[i];
            if (c == '\x1b')
            {
                if (i + 1 < text.Length && text[i + 1] == '[')
                {
                    int j = i + 2;
                    while (j < text.Length && !char.IsAsciiLetter(text[j])) j++;
                    i = j + 1;
                    continue;
                }
                if (i + 1 < text.Length && text[i + 1] == ']')
                {
                    int j = i + 2;
                    while (j < text.Length && text[j] != '\x07') j++;
                    i = j + 1;
                    continue;
                }
                i += 2;
                continue;
            }
            if (c < ' ' && c != '\n' && c != '\r' && c != '\t') { i++; continue; }
            result.Append(c);
            i++;
        }
        return result.ToString();
    }
}
